crates/hotshot/src/traits/networking/push_cdn_network.rs

```diff
@@ -0,0 +1,424 @@
+    ///
+    /// # Errors
+    /// - If we fail to serialize the message
+    /// - If we fail to send the broadcast message.
+    async fn da_broadcast_message(
+        &self,
+        message: Message<TYPES>,
+        _recipients: BTreeSet<TYPES::SignatureKey>,
+    ) -> Result<(), NetworkError> {
+        self.broadcast_message(message, Topic::DA).await
+    }
+
+    /// Send a direct message to a node with a particular key. Does not retry.
+    ///
+    /// - If we fail to serialize the message
+    /// - If we fail to send the direct message
+    async fn direct_message(
+        &self,
+        message: Message<TYPES>,
+        recipient: TYPES::SignatureKey,
+    ) -> Result<(), NetworkError> {
+        // Bincode the message
+        let serialized_message = match bincode_opts().serialize(&message) {
+            Ok(serialized) => serialized,
+            Err(e) => {
+                warn!("Failed to serialize message: {}", e);
+                return Err(NetworkError::FailedToSerialize { source: e });
+            }
+        };
+
+        // Send the message
+        // TODO: check if we need to print this error
+        if self
+            .0
+            .send_direct_message(&WrappedSignatureKey(recipient), serialized_message)
+            .await
+            .is_err()
+        {
+            return Err(NetworkError::CouldNotDeliver);
+        };
+
+        Ok(())
+    }
+
+    /// Receive a message. Is agnostic over `transmit_type`, which has an issue
+    /// to be removed anyway.
+    ///
+    /// # Errors
+    /// - If we fail to receive messages. Will trigger a retry automatically.
+    async fn recv_msgs(&self) -> Result<Vec<Message<TYPES>>, NetworkError> {
+        // Receive a message
+        let message = self.0.receive_message().await;
+
+        // If it was an error, wait a bit and retry
+        let message = match message {
+            Ok(message) => message,
+            Err(error) => {
+                async_sleep(Duration::from_millis(100)).await;
+                error!("failed to receive message: {error}");
+                return Err(NetworkError::PushCdnNetwork {
+                    source: PushCdnNetworkError::FailedToReceive,
+                });
+            }
+        };
+
+        // Figure out if it's the right type
+        if let PushCdnMessage::Direct(data) = message {
+            // Deserialize it
+            let result: Message<TYPES> = bincode_opts()
+                .deserialize(&data.message)
+                .map_err(|e| NetworkError::FailedToSerialize { source: e })?;
+
+            // Return it
+            return Ok(vec![result]);
+        }
+
+        if let PushCdnMessage::Broadcast(data) = message {
+            // Deserialize it
+            let result: Message<TYPES> = bincode_opts()
+                .deserialize(&data.message)
+                .map_err(|e| NetworkError::FailedToSerialize { source: e })?;
+            return Ok(vec![result]);
+        }
```

Comment:
Can this be combined?  We no longer need to split direct and broadcast on the receiver side in Hotshot.  I'm guessing this was a result of their existing `TransmitType` when you were implementing this.

Reply:
Yep, totally correct. Think you had removed that after I wrote this

Reply:
Okay, on another run through: `data` is actually a direct or broadcast-specific message (needed to add recipient or topic metadata embedded in the message), which means they cannot be thrown in a match arm or anything. I can do some deduplication though. 

The other option is to have the broker remove the metadata and just send to the user as a `Data` message directly. I was originally doing this but wanted to reduce broker (re)allocations and cycles.